tests/benchmarks/bench_assembly.cpp

```diff
@@ -0,0 +1,389 @@
+   {
+      a.Assemble();
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs();
+   }
+
+   void benchmark() override
+   {
+      cg.Mult(B,X);
+      MFEM_DEVICE_SYNC;
+      mdofs += MDofs() * cg.GetNumIterations();
+   }
+};
+
+/// Bake-off Problems (BPs)
+#define BakeOff_Problem(assembly,i,Kernel,VDIM,p_eq_q)\
+static void SetupBP##i##assembly(bm::State &state){\
+   const int dim = 3;\
+   const int p = state.range(1);\
+   const int target_dofs = state.range(0);\
+   const int elem_dofs = pow(p+1, dim);\
+   const int N = pow(target_dofs / elem_dofs, 1.0/dim) + 1;\
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(AssemblyLevel::assembly, p, N);\
+   if ( !ker.is_runnable() ) { state.SkipWithError("MAX_MEM"); }\
+   while (state.KeepRunning()) { ker.setup(); }\
+   state.counters["MDof/s"] = bm::Counter(ker.SumMdofs(), bm::Counter::kIsRate);\
+   state.counters["Dofs"] = bm::Counter(ker.dofs, bm::Counter::kDefaults);\
+   state.counters["Order"] = bm::Counter(ker.p);\
+   state.counters["Assembly"] = (int)AssemblyLevel::assembly;}\
+BENCHMARK(SetupBP##i##assembly)->ArgsProduct({\
+      benchmark::CreateRange(1024, max_dofs, /*step=*/2),\
+      benchmark::CreateDenseRange(1, max_order, /*step=*/1)\
+    })->Unit(bm::kMillisecond);\
+static void BP##i##assembly(bm::State &state){\
+   const int dim = 3;\
+   const int p = state.range(1);\
+   const int target_dofs = state.range(0);\
+   const int elem_dofs = pow(p+1, dim);\
+   const int N = pow(target_dofs / elem_dofs, 1.0/dim) + 1;\
+   Problem<Kernel##Integrator,VDIM,p_eq_q> ker(AssemblyLevel::assembly, p, N);\
+   if ( !ker.is_runnable() ) { state.SkipWithError("MAX_MEM"); }\
+   while (state.KeepRunning()) { ker.benchmark(); }\
+   state.counters["MDof/s"] = bm::Counter(ker.SumMdofs(), bm::Counter::kIsRate);\
+   state.counters["Dofs"] = bm::Counter(ker.dofs, bm::Counter::kDefaults);\
+   state.counters["Order"] = bm::Counter(ker.p);\
+   state.counters["Assembly"] = (int)AssemblyLevel::assembly;}\
+BENCHMARK(BP##i##assembly)->ArgsProduct({\
+      benchmark::CreateRange(1024, max_dofs, /*step=*/2),\
+      benchmark::CreateDenseRange(1, max_order, /*step=*/1)\
+    })->Unit(bm::kMillisecond);
+
+// PARTIAL:
+/// BP1: scalar PCG with mass matrix, q=p+2
+BakeOff_Problem(PARTIAL,1,Mass,1,false)
+
+/// BP2: vector PCG with mass matrix, q=p+2
+BakeOff_Problem(PARTIAL,2,VectorMass,3,false)
+
+/// BP3: scalar PCG with stiffness matrix, q=p+2
+BakeOff_Problem(PARTIAL,3,Diffusion,1,false)
+
+/// BP4: vector PCG with stiffness matrix, q=p+2
+BakeOff_Problem(PARTIAL,4,VectorDiffusion,3,false)
+
+/// BP5: scalar PCG with stiffness matrix, q=p+1
+BakeOff_Problem(PARTIAL,5,Diffusion,1,true)
+
+/// BP6: vector PCG with stiffness matrix, q=p+1
+BakeOff_Problem(PARTIAL,6,VectorDiffusion,3,true)
+
+// ELEMENT:
+/// BP1: scalar PCG with mass matrix, q=p+2
+BakeOff_Problem(ELEMENT,1,Mass,1,false)
+
+/// BP2: vector PCG with mass matrix, q=p+2
+// BakeOff_Problem(ELEMENT,2,VectorMass,3,false)
```

Comment:
Why are several benchmarks left as commented?

Reply:
Because they run out of memory, I'm open to a better solution if you have propositions.

Reply:
Some of them are also not yet implemented.

Reply:
I added some "Not yet implemented" comments, tell me if you prefer me to remove the lines instead.